lib/pbench/agent/utils.py

```diff
@@ -390,3 +393,60 @@ def collect_local_info(pbench_bin):
         hostdata[arg] = cp.stdout.strip() if cp.stdout is not None else ""
 
     return (version, seqno, sha1, hostdata)
+
+
+if os.environ.get("_PBENCH_UNIT_TESTS"):
+
+    class LocalRemoteHost:
+        """Mock out behavior for legacy unit tests.
+        """
+
+        def __init__(self):
+            self._local_names = frozenset(
+                [
+                    "127.0.0.1",
+                    "localhost",
+                    "172.16.42.42",
+                    "testhost.example.com",
+                    "testhost",
+                ]
+            )
+
+        def is_remote(self, host_name):
+            return host_name not in self._local_names
+
+
+else:
+
+    class LocalRemoteHost:
+        """Dynamically build a list of all the local IP addresses so that a simple
+        simple query interface provides the answer to the question, "is the given
+        host name or IP address remote?"
+        """
+
+        def __init__(self):
+            """Build up a list of local IP addresses from all the interfaces.
+            """
+            local_ips_l = []
+            for adapter in ifaddr.get_adapters():
+                for ip in adapter.ips:
+                    local_ips_l.append(ip.ip[0] if isinstance(ip.ip, tuple) else ip.ip)
+            self.local_ips = frozenset(local_ips_l)
+
+        def is_remote(self, host_name):
+            """Returns True if the given host name (or IP address) is remote else
+            returns False.
+            """
+            ips_l = []
+            try:
+                ip_addrs = socket.getaddrinfo(host_name, None)
+                for addrinfo in ip_addrs:
+                    ips_l.append(addrinfo[4][0])
+            except socket.gaierror:
+                try:
+                    ipaddress.ip_address(host_name)
+                except ValueError:
+                    return True
+                ips_l.append(host_name)
+            ips = frozenset(ips_l)
+            return self.local_ips.isdisjoint(ips)
```

Comment:
I think that we want `is_remote()` to return `True` unless we're sure that `ssh` would be hitting the local host.  So, if _any_ of the translations aren't in the local set, then I think we should probably return `True`.  So, `isdisjoint()` isn't the right operation -- we need something more like `issubset()`.

(Of course, if we can limit our comparison to a single target address, then we can just use `in` and dispense with producing `frozenset`'s here....)

Reply:
Just to throw this out here for "posterity" (whomever that may turn out to be), I needed this PR's basic logic but, partially by way of fixing some of the weaknesses here (especially in IPv6), and partially just because I felt like it, I put my own spin on it... including reversing the test. (Mostly because the usage in the code ends up being "is local" vs "not is remote".)

I haven't done unit testing, nor tried to accommodate the legacy test environment variable, but I think there are two important aspects to contribute to this PR:

1. Proper handling of IPv6 comparison using `ipaddress` rather than string matches;
2. Not trying to do some sort of full set match but rather looking only for proof that (at least) one translation of the parameter hostname is a match for a `localhost` representation. (Webb may have a point about making sure that *all* host representations are in `localhost` but I'm not at all convinced the sort of split routing that implies is feasible and I didn't try to deal with it...)

__NOTE__ that I'm ignoring all the translations that aren't INET or INET6 family as that's all we care about. I'm also ignoring anything with a non-zero IPv6 scope_id on the principle that they're probably not interesting or useful, as well as the pragmatic aspect that these will include the `%scope` suffix on the text IPv6 representation, which `ipaddress.ip_address()` can't parse. It's not entirely impossible that we could have a captive subnet on one interface that's the only path to our remote nodes, and I'm not certain whether that's realistic; but for now I decided not to bother with it.

In short, I'm throwing this out as a follow-up comment, as sample, and as a prototype, not as the answer to life, the universe, and everything.

```
class HostComparator:
    """
    Determine whether a given hostname or IP is an alias for localhost.
    """

    def __init__(self):
        """
        Build a list of local IP addresses for all interfaces
        """
        ips = []
        for adapter in ifaddr.get_adapters():
            for ip in adapter.ips:
                if isinstance(ip.ip, tuple):
                    if len(ip.ip) > 2 and ip.ip[2] == 0:
                        # This is an IPv6 address: use it if the scope ID is 0
                        # (That is, it's routable through any interface)
                        ips.append(ipaddress.ip_address(ip.ip[0]))
                else:
                    ips.append(ipaddress.ip_address(ip.ip))
        self.aliases = frozenset(ips)

    def is_local(self, host: str) -> bool:
        """
        Determine whether a given hostname is an alias for the local host; in
        other words, whether it matches a canonical representation of one of
        the routable addresses reported by `ifaddr`.

        Args:
            host: A hostname or IP address to check

        Returns:
            True if `host` matches a routable local host address
        """
        try:
            infos = socket.getaddrinfo(host, None)

            # getaddrinfo returns 5-tuples of names for the host in the format
            # (family, type, proto, name, (address, port[, flowinfo, scopeid]))
            # where flowinfo and scopeid are present only for IPv6. We want to
            # capture all IPv4 addresses plus all IPv6 addresses without a
            # scope ID. If any of the resulting addresses match against the
            # "localhost" alias set, return true.
            for info in infos:
                proto = info[0]
                addr = info[4]
                if proto == socket.AF_INET or (proto == socket.AF_INET6 and addr[3] == 0):
                    ip = ipaddress.ip_address(addr[0])
                    if ip in self.aliases:
                        return True
        except socket.gaierror:
            # If address lookup fails, check whether the hostname is legal
            try:
                addr = ipaddress.ip_address(host)
                return addr in self.aliases
            except ValueError:
                return False
```